### PR TITLE
[SPARK-38760][PYTHON][SQL][SS] Add DataFrame.observe(str, ...) for Structured Streaming in PySpark

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR adds the support of Structured Streaming at `DataFrame.observe` in PySpark. This is the same support with the Scala side.

### Why are the changes needed?

For SS users in PySpark to easily monitor their streaming queries.

### Does this PR introduce _any_ user-facing change?

Yes. After this PR, PySpark users can do:

```python
from pyspark.sql.functions import count, col, sum, lit
from pyspark.sql.streaming import StreamingQueryListener

class MyListener(StreamingQueryListener):
    def onQueryStarted(self, event):
        pass
    def onQueryProgress(self, event):
        row = event.progress.observedMetrics.get("metric")
        # Do something with the metric, e.g.) row.cnt and row.sum
        print("Count: %s" % row.cnt)
    def onQueryTerminated(self, event):
        pass

spark.streams.addListener(MyListener())

df = spark.readStream.format("rate").option("rowsPerSecond", 10).load()
observed_df = df.observe("metric", count(lit(1)).alias("cnt"), sum(col("value")).alias("sum"))
observed_df.writeStream.format("noop").queryName("test").start()
```

### How was this patch tested?

Manually tested with the example above, and unit test was added.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @HeartSaVioR @alex-balikov @hvanhovell @zsxwing, this PR is ready for a look. Please take a look when you find some time.

**Comment by HyukjinKwon** [comment]
cc @belifer too

File: sql/core/src/main/scala/org/apache/spark/sql/api/python/PythonSQLUtils.scala [hunk 1]
```diff
@@ -37,7 +38,10 @@ import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.types.DataType
 
 private[sql] object PythonSQLUtils extends Logging {
-  private lazy val rowPickler = new Pickler(true, false)
+  private lazy val internalrowPickler = {
```
Comment (xinrong-databricks) [inline sql/core/src/main/scala/org/apache/spark/sql/api/python/PythonSQLUtils.scala:41]: nit: internal**R**owPickler

**Review by HeartSaVioR — APPROVED** [review]
+1 assume I could delegate Python specific review to @xinrong-databricks and this has been manually tested with multi-workers environment as well.

**Comment by HyukjinKwon** [comment]
Thanks guys.

Merged to master.

**Pull request closed without merging**